demo/bulk_insert/bulk_insert.py

```diff
@@ -1,281 +1,339 @@
+
+    def to_binary(self):
+        return self.packed_header + b''.join(self.binary_entities)
+
+# Handler class for processing label csv files.
+class Label(EntityFile):
+    def __init__(self, infile):
+        super(Label, self).__init__(infile)
+        expected_col_count = self.process_header()
+        self.process_entities(expected_col_count)
+        self.infile.close()
+
+    def process_header(self):
+        # Header format:
+        # node identifier (which may be a property key), then all other property keys
+        header = next(self.reader)
+        expected_col_count = len(header)
+        # If identifier field begins with an underscore, don't add it as a property.
+        if header[0][0] == '_':
+            self.prop_offset = 1
+        self.packed_header = self.pack_header(header)
+        self.binary_size += len(self.packed_header)
+        #  self.binary_entities.append(self.packed_header)
+        return expected_col_count
+
+    def process_entities(self, expected_col_count):
+        global NODE_DICT
+        global TOP_NODE_ID
+        global QUERY_BUF
+
         for row in self.reader:
-            # Raise an exception if the wrong number of columns are found
-            if len(row) != expected_col_count:
-                raise CSVError ("%s:%d Expected %d columns, encountered %d ('%s')"
-                                % (self.csvfile, self.reader.line_num, expected_col_count, len(row), ','.join(row)))
-            if (row[-1] == ''):
-                raise CSVError ("%s:%d Dangling comma in input. ('%s')"
-                                % (self.csvfile, self.reader.line_num, ','.join(row)))
-        # Subtract 1 from each file's entity count to compensate for the header.
-        self.total_entities = self.reader.line_num - 1
-
-    def token_count(self):
-        # Labels have a token for name, attribute count, pending insertion count, plus N tokens for the individual attributes.
-        return 3 + self.attribute_count
-
-    def unroll(self):
-        return [self.name, self.pending_inserts, self.attribute_count] + self.attributes
-
-    def print_insertion_start(self):
-        print('Inserting Label "%s" - %d nodes' % (self.name, self.total_entities))
-
-# A RelationDescriptor consists of a name string and a relationship count.
-# As a query string, a RelationDescriptor is printed in the format:
-# [relation name, relation count]
-class RelationDescriptor(Descriptor):
-    def __init__(self, csvfile):
-        Descriptor.__init__(self, csvfile)
-        self.attribute_count = 2
-        self.validate_csv()
-        self.csvfile.seek(0)
-
-    def validate_csv(self):
+            self.validate_row(expected_col_count, row)
+            # Add identifier->ID pair to dictionary if we are building relations
+            if NODE_DICT is not None:
+                if row[0] in NODE_DICT:
+                    print("Node identifier '%s' was used multiple times - second occurrence at %s:%d"
```

Comment:
Do we really want to proceed in this case?

Reply:
:+1: